### PR TITLE
Normalize extras in lockfile 

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

Previously, when we locked something like `flask[dotenv]`, we created two separate distributions in the lockfile: one for `flask`, which included the base dependencies, and one for `flask[dotenv]`, which included the base dependencies _and_ the `dotenv` dependencies. This was easy to implement, but it meant that we were duplicating all of the distribution files for every extra, and duplicating all of the base dependencies for every extra.

This PR normalizes the data such that we now have one entry per distribution (i.e., `ExtraName` was removed from `DistributionId`), with an optional dependencies table with an entry per extra, like:

```toml
[[distribution]]
name = "project"
version = "0.1.0"
source = "editable+file://[TEMP_DIR]/"
sdist = { url = "file://[TEMP_DIR]/" }

[[distribution.dependencies]]
name = "anyio"
version = "3.7.0"
source = "registry+https://pypi.org/simple"

[distribution.optional-dependencies]

[[distribution.optional-dependencies.test]]
name = "iniconfig"
version = "2.0.0"
source = "registry+https://pypi.org/simple"
```

This requires a bit more work upfront, because we now need to merge multiple packages from the `PetGraph` representation when creating the lockfile.

Closes https://github.com/astral-sh/uv/issues/3916.

### CODE REVIEW

File: crates/uv-resolver/src/lock.rs [hunk 1]
```diff
@@ -504,21 +538,17 @@ impl Distribution {
...
     pub(crate) name: PackageName,
     pub(crate) version: Version,
-    #[serde(skip_serializing_if = "Option::is_none")]
-    pub(crate) extra: Option<ExtraName>,
     pub(crate) source: Source,
 }
 
 impl DistributionId {
-    fn from_annotated_dist(annotated_dist: &AnnotatedDist) -> DistributionId {
+    pub(crate) fn from_annotated_dist(annotated_dist: &AnnotatedDist) -> DistributionId {
```
Comment (charliermarsh) [inline crates/uv-resolver/src/lock.rs:545]: Not thrilled that I'm making so many things `pub(crate)` here. Should I change `ResolutionGraph::to_lock` into a `impl TryFrom<ResolutionGraph` for `Lock`?

**Comment by charliermarsh** [comment]
@BurntSushi - Ignoring the code, curious if you prefer this representation?

File: crates/uv-resolver/src/resolution/graph.rs [hunk 2]
```diff
@@ -442,18 +443,42 @@ impl ResolutionGraph {
...
+                continue;
+            }
+
             let mut locked_dist = lock::Distribution::from_annotated_dist(dist)?;
             for neighbor in self.petgraph.neighbors(node_index) {
                 let dependency_dist = &self.petgraph[neighbor];
                 locked_dist.add_dependency(dependency_dist);
             }
-            locked_dists.push(locked_dist);
+            if let Some(locked_dist) = locked_dists.insert(locked_dist.id.clone(), locked_dist) {
```
Comment (ibraheemdev) [inline crates/uv-resolver/src/resolution/graph.rs:461]: Why could the previous code do an unconditional `push` here?

**Comment by charliermarsh** [comment]
I think I will leave the representation as-is for now because it closely mirrors the `pyproject.toml` schema, where you have an `optional-dependencies` map that's keyed on extra name. That was intentional, because I'm hoping to make it possible to reify the distribution metadata from the lockfile in the future. It's a very good question though.

File: crates/uv-resolver/src/lock.rs (hunk 1 above)
Comment (charliermarsh) [reply crates/uv-resolver/src/lock.rs:545]: Yeah I'm also not totally convinced that `ResolutionGraph` will be necessary in the long run.

**Pull request merged**